src/dotty/tools/dotc/core/ConstraintHandling.scala

```diff
@@ -0,0 +1,256 @@
+ *          H | H
+ *
+ *  (the idea is that a parameter Q in a H context is guaranteed to be a supertype of P).
+ *
+ * "P >: Q in C" means: There is a constraint P >: L[Q],
+ *     where L is the multi-hole context given by:
+ *
+ *      L = []
+ *          L | T
+ *          T | L
+ *          L & L
+ */
+trait ConstraintHandling {
+  
+  implicit val ctx: Context
+  
+  protected def isSubType(tp1: Type, tp2: Type): Boolean
+  
+  val state: TyperState
+  import state.constraint
+  
+  private var addConstraintInvocations = 0
+
+  /** If the constraint is frozen we cannot add new bounds to the constraint. */
+  protected var frozenConstraint = false
+   
+  private def addOneBound(param: PolyParam, bound: Type, isUpper: Boolean): Boolean =
+    !constraint.contains(param) || {
+      val c1 = constraint.narrowBound(param, bound, isUpper)
+      (c1 eq constraint) || {
+        constraint = c1
+        val TypeBounds(lo, hi) = constraint.entry(param)
+        isSubType(lo, hi)
+      }
+    }
+
+  protected def addUpperBound(param: PolyParam, bound: Type): Boolean = {
+    def description = i"constraint $param <: $bound to\n$constraint"
+    if (bound.isRef(defn.NothingClass) && ctx.typerState.isGlobalCommittable) {
+      def msg = s"!!! instantiated to Nothing: $param, constraint = ${constraint.show}"
+      if (Config.flagInstantiationToNothing) assert(false, msg)
+      else ctx.log(msg)
+    }
+    constr.println(i"adding $description")
+    val lower = constraint.lower(param)
+    val res =
+      addOneBound(param, bound, isUpper = true) &&
+      lower.forall(addOneBound(_, bound, isUpper = true))
+    constr.println(i"added $description = $res")
+    res
+  }
+    
+  protected def addLowerBound(param: PolyParam, bound: Type): Boolean = {
+    def description = i"constraint $param >: $bound to\n$constraint"
+    constr.println(i"adding $description")
+    val upper = constraint.upper(param)
+    val res =
+      addOneBound(param, bound, isUpper = false) &&
+      upper.forall(addOneBound(_, bound, isUpper = false))
+    constr.println(i"added $description = $res")
+    res
+  }
+ 
+  protected def addLess(p1: PolyParam, p2: PolyParam): Boolean = {
+    def description = i"ordering $p1 <: $p2 to\n$constraint"
+    val res =
+      if (constraint.isLess(p2, p1)) unify(p2, p1) 
+        // !!! this is direction dependent - unify(p1, p2) makes i94-nada loop forever.
+        //     Need to investigate why this is the case. 
+        //     The symptom is that we get a subtyping constraint of the form P { ... } <: P
+      else {
+        val down1 = p1 :: constraint.exclusiveLower(p1, p2)
+        val up2 = p2 :: constraint.exclusiveUpper(p2, p1)
+        val lo1 = constraint.nonParamBounds(p1).lo
+        val hi2 = constraint.nonParamBounds(p2).hi
+        constr.println(i"adding $description down1 = $down1, up2 = $up2")
+        constraint = constraint.addLess(p1, p2)
+        down1.forall(addOneBound(_, hi2, isUpper = true)) &&
```

Comment:
Could the difference between `Constraint` and `ConstraintHandler` be written down explicitly somewhere? I don't understand why this line and the next are not part of `Constraint#addLess`.